File: 07-BACKTRACKING/Go/combination-sum.go
package main

// https://leetcode.com/problems/combination-sum/
func combinationSum(candidates []int, target int) [][]int {
	var res [][]int

	var backtrack func(index int, curr []int, total int)
	backtrack = func(index int, curr []int, total int) {
		if total == target {
			currCopy := make([]int, len(curr))
			copy(currCopy, curr)
			res = append(res, currCopy)
			return
		}

		if total > target || index >= len(candidates) {
			return
		}

		curr = append(curr, candidates[index])
		backtrack(index, curr, total+candidates[index])
		curr = curr[:len(curr)-1]
		backtrack(index+1, curr, total)
	}

	backtrack(0, []int{}, 0)
	return res
}

File: 07-BACKTRACKING/PHP/combination-sum.php
<?php

// https://leetcode.com/problems/combination-sum/
class Solution
{
    private array $candidates = [];
    private int $target = 0;
    private array $res = [];

    /**
     * @param Integer[] $candidates
     * @param Integer $target
     * @return Integer[][]
     */
    function combinationSum(array $candidates, int $target): array
    {
        $this->target = $target;
        $this->candidates = $candidates;

        $curr = [];
        $this->dfs(0, $curr, 0);

        return $this->res;
    }

    private function dfs(int $index, array &$curr, int $total): void
    {
        if ($total == $this->target) {
            $this->res[] = $curr;
            return;
        }

        if ($total > $this->target || $index >= count($this->candidates)) {
            return;
        }

        $curr[] = $this->candidates[$index];
        $this->dfs($index, $curr, $total + $this->candidates[$index]);

        array_pop($curr);
        $this->dfs($index + 1, $curr, $total);
    }
}
